- `Import Libraries`

In [31]:
import multiprocessing
import os
import time

- `Number of Cores in System`

In [4]:
print("Number of CPU Cores: ", multiprocessing.cpu_count())

Number of CPU Cores:  4


- `Process : To Invoke Individual Process & Do Some Task`
- `Join Function : Used to Terminate Process`

In [25]:
def display_fruit(fruit_name):
    print("ProcessId: {}, Fruit Name: {}".format(os.getpid(), fruit_name))

if __name__ == "__main__":
    fruits_list = ["Apple", "Banana", "Grapes", "Mango"]
    print("Main Process ID : {}".format(os.getpid()))

    # Creating Child Process
    p1 = multiprocessing.Process(target=display_fruit, args=("Apple",))
    p2 = multiprocessing.Process(target=display_fruit, args=("Mango",))

    # Starting Child Processes
    p1.start()
    p2.start()
    print("--"*15)
    print("Before Process Termination")
    print("--"*15)    
    print("Is Child Process[p1] Alive: {}".format(p1.is_alive()))
    print("Is Child Process[p2] Alive: {}".format(p2.is_alive()))

    # Terminates Child Process
    p1.join()
    p2.join()
    print("--"*15)
    print("After Process Termination")
    print("--"*15)
    print("Is Child Process[p1] Alive: {}".format(p1.is_alive()))
    print("Is Child Process[p2] Alive: {}".format(p2.is_alive()))    

Main Process ID : 4907
ProcessId: 5056, Fruit Name: Apple
ProcessId: 5057, Fruit Name: Mango
------------------------------
Before Process Termination
------------------------------
Is Child Process[p1] Alive: True
Is Child Process[p1] Alive: True
------------------------------
After Process Termination
------------------------------
Is Child Process[p1] Alive: False
Is Child Process[p1] Alive: False


- `Queue Function : Used to Store data like First-In-First-Out object & can be shared between process`

In [29]:
def display_fruit(fruit_name):
    print("ProcessId: {}, Fruit Name: {}".format(os.getpid(), fruit_name))

if __name__ == "__main__":
    fruits_list = ["Apple", "Banana", "Grapes", "Mango"]
    counter = 0
    # Initialising Queue Object
    fruits_queue = multiprocessing.Queue()
    print("--"*20)
    print("Inserting/Pushing Fruits Into Queue Object")
    print("--"*20)
    for fruit in fruits_list:
        print(f"Fruit Number : {counter}, Fruit Name {fruit}")
        fruits_queue.put(fruit)
        counter += 1
    print("--"*20)
    print("Taking/Popping Out Fruits From Queue Object")
    print("--"*20)    
    counter = 0
    while not fruits_queue.empty():
        print(f"Fruit Number : {counter}, Fruit Name {fruits_queue.get()}")
        counter += 1

----------------------------------------
Inserting/Pushing Fruits Into Queue Object
----------------------------------------
Fruit Number : 0, Fruit Name Apple
Fruit Number : 1, Fruit Name Banana
Fruit Number : 2, Fruit Name Grapes
Fruit Number : 3, Fruit Name Mango
----------------------------------------
Taking/Popping Out Fruits From Queue Object
----------------------------------------
Fruit Number : 0, Fruit Name Apple
Fruit Number : 1, Fruit Name Banana
Fruit Number : 2, Fruit Name Grapes
Fruit Number : 3, Fruit Name Mango


- `Pool : Used for parallel execution of a function of given input values`
- `where input data shared across multiple processes created by pool i.e., Data Parallelism(Like each input value is assigned to each process)`

In [36]:
def display_fruit(fruit_name):
    if fruit_name == "Apple":
        time.sleep(5)
    if fruit_name == "Banana":
        time.sleep(10)
    if fruit_name == "Grapes":
        time.sleep(15)
    if fruit_name == "Mango":
        time.sleep(20)                        
    print("Time: %s, ProcessId: %s, Fruit Name: %s" % (time.strftime("%Y-%m-%d %H:%M:%S"), os.getpid(), fruit_name))

if __name__ == "__main__":
    fruits_list = ["Apple", "Banana", "Grapes", "Mango"]
    print("Main Process ID : {}".format(os.getpid()))
    print("Current Time Before Starting Pool Process %s " % (time.strftime("%Y-%m-%d %H:%M:%S")))
    # Creating Pool of Processes
    pool = multiprocessing.Pool(len(fruits_list))
    # Starting Processes in Pool
    # Note : All Process starts at same time, but its gets completed based on its condition time sleep
    pool.map(display_fruit, fruits_list)

Main Process ID : 4907
Current Time Before Starting Pool Process 2022-02-03 17:57:12 
Time: 2022-02-03 17:57:17, ProcessId: 5140, Fruit Name: Apple
Time: 2022-02-03 17:57:22, ProcessId: 5141, Fruit Name: Banana
Time: 2022-02-03 17:57:27, ProcessId: 5142, Fruit Name: Grapes
Time: 2022-02-03 17:57:32, ProcessId: 5143, Fruit Name: Mango


> `Observation`
- From Current Time [`2022-02-03 17:57:12`], ProcessId: 5140, Fruit Name: Apple takes 5 seconds & completes on [`2022-02-03 17:57:17`]
- From Current Time [`2022-02-03 17:57:12`], ProcessId: 5141, Fruit Name: Apple takes 10 seconds & completes on [`2022-02-03 17:57:22`]
- From Current Time [`2022-02-03 17:57:12`], ProcessId: 5142, Fruit Name: Apple takes 15 seconds & completes on [`2022-02-03 17:57:27`]
- From Current Time [`2022-02-03 17:57:12`], ProcessId: 5143, Fruit Name: Apple takes 20 seconds & completes on [`2022-02-03 17:57:32`]